File: src/abstract_factory/enchanted/EnchantedMazeFactory.java
package abstract_factory.enchanted;

import abstract_factory.base.Door;
import abstract_factory.base.MazeFactory;
import abstract_factory.base.Room;

public class EnchantedMazeFactory extends MazeFactory {

    @Override
    public Room makeRoom(int n) {
        return new EnchantedRoom(n, castSpell());
    }

    @Override
    public Door makeDoor(Room r1, Room r2) {
        return new DoorNeedingSpell(r1, r2);
    }

    private Spell castSpell() {
        return new Spell();
    }

}

File: src/abstract_factory_kotlin/enchanted/EnchantedMazeFactory.kt
package abstract_factory_kotlin.enchanted

import abstract_factory_kotlin.base.MazeFactory
import base.maze.Room

class EnchantedMazeFactory : MazeFactory() {
    override fun makeRoom(n: Int) = EnchantedRoom(n, castSpell())
    override fun makeDoor(r1: Room, r2: Room) = DoorNeedingSpell(r1, r2)
    private fun castSpell() = Spell()
}
